### PR TITLE
[IN-417] V2 Claim Unregistered User Endpoint

### PR DESCRIPTION
**Pull request by mfraezz**
## Purpose
Allow Ember apps to let unregistered users be claimed

## Changes
* Add endpoint
* Add tests

## QA Notes
No UI components, API testing would duplicate TCI

## Side Effects
None expected

## Ticket
[[IN-417]](https://openscience.atlassian.net/browse/IN-417)

### CODE REVIEW

File: api/users/views.py [hunk 1]
```diff
@@ -436,3 +442,77 @@ def perform_destroy(self, instance):
...
+        rely upon a flask context and placed in utils (or elsewhere).
+
+        bool registered : Indicates which sender to call
+        args, kwargs    : Passed through to mail helpers
+        return          : None
+        """
+        from website.app import app
+        from website.routes import make_url_map
+        try:
+            make_url_map(app)
```
Comment (sloria) [inline api/users/views.py]: Just curious: is there a situation where we would hit this method without the routes being initialized? I guess I'm not sure why this is necessary.

File: api_tests/users/views/test_user_claim.py [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+            }
+        }
+        _id = kwargs.pop('id', None)
+        if _id:
+            payload['data']['id'] = _id
+        if kwargs:
+            payload['data']['attributes'] = kwargs
+        return payload
+
+    def test_unacceptable_methods(self, app, url, claimer, unreg_user, project):
```
Comment (sloria) [inline api_tests/users/views/test_user_claim.py]: Not sure if this test is necessary--it's mostly testing the framework code.

File: api_tests/users/views/test_user_claim.py [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+        _url = url.format(project.creator._id)
+        payload = self.payload(email=project.creator.email, id=project._id)
+        res = app.post_json_api(
+            _url,
+            payload,
+            expect_errors=True
+        )
+        assert res.status_code == 401
+
+    @mock.patch('website.project.views.contributor.mails.send_mail')
```
Comment (sloria) [inline api_tests/users/views/test_user_claim.py]: Nit: This could be made into a fixture

```python
@pytest.fixture
def mock_mail(self):
    with mock.patch('website.project.views.contributor.mails.send_mail') as patch:
        yield patch
```

File: website/project/views/contributor.py [hunk 4]
```diff
@@ -361,6 +361,7 @@ def project_remove_contributor(auth, **kwargs):
     return redirect_url
 
 
+# TODO: consider moving this into utils
```
Comment (sloria) [inline website/project/views/contributor.py]: 👍

File: api/users/permissions.py [hunk 5]
```diff
@@ -34,3 +34,16 @@ def has_object_permission(self, request, view, obj):
...
+    Allows no one to attempt to claim a registered user.
+    """
+    def has_permission(self, request, view):
+        claimed_user = view.get_user(check_permissions=False)
+        assert isinstance(claimed_user, OSFUser), 'obj must be a User, got {}'.format(claimed_user)
+        return not claimed_user.is_registered
+
+    def has_object_permission(self, request, view, obj):
+        assert isinstance(obj, OSFUser), 'obj must be a User, got {}'.format(obj)
+        return not obj.is_registered
```
Comment (alexschiller) [inline api/users/permissions.py]: Disabled accounts get set is_registered = False so they would get past here. Might that cause problems?

Aside, the "claim user" dialogue shows up for disabled accounts, but breaks when attempted, so there are definitely issues with that workflow that are unrelated to this PR.

File: api/users/permissions.py [hunk 6]
```diff
@@ -34,3 +34,16 @@ def has_object_permission(self, request, view, obj):
             return True
         else:
             return obj['self']._id == request_user._id
+
+class ClaimUserPermission(permissions.BasePermission):
+    """ Allows anyone to attempt to claim an unregistered user.
+    Allows no one to attempt to claim a registered user.
+    """
+    def has_permission(self, request, view):
+        claimed_user = view.get_user(check_permissions=False)
```
[the remark this answers is not in the document]
Comment (mfraezz) [reply api/users/permissions.py]: The user that `view.get_user` returns is guaranteed to not be the `request.user` in all success cases.

File: api/users/views.py [hunk 7]
```diff
@@ -436,3 +442,77 @@ def perform_destroy(self, instance):
...
+        claimer = request.user
+        email = (request.data.get('email', None) or '').lower().strip()
+        try:
+            record_id = request.data.get('id', '').lower().strip()
+        except AttributeError:
+            # Parsed as None if not present
+            raise ValidationError('Must specify record "id".')
+        claimed_user = self.get_user(check_permissions=True)  # Ensures claimability
+        try:
+            record_referent = Guid.objects.get(_id=record_id).referent
```
[the remark this answers is not in the document]
Comment (mfraezz) [reply api/users/views.py]: A user guid should never be passed here, but `Node` and `Preprint` guids are permissible.

File: api/users/views.py (hunk 1 above)
Comment (mfraezz) [reply api/users/views.py]: It's outside of a flask context, so the first time it's hit (indicated by test behavior, at least).

File: api_tests/users/views/test_user_claim.py (hunk 2 above)
Comment (mfraezz) [reply api_tests/users/views/test_user_claim.py]: This test is more of a canary against accidental edits than an assurance that DRF behaves as expected, but I can remove if you want.

File: api/users/permissions.py (hunk 5 above)
Comment (mfraezz) [reply api/users/permissions.py]: It seems like disabled accounts shouldn't have `is_registered` set to `False`. A `disabled` `NonNaiveDateTimeField` is probably the more idiomatic way to do that... and we've already got one. The refactor to tear out any use of `is_registered` in that flow is out of scope for this.

I can add a check for `is_claimed` here as well, but that might have other side effects, based on [the TODO above `is_claimed`](https://github.com/CenterForOpenScience/osf.io/blob/65ded1a/osf/models/user.py#L177-L180).

File: api_tests/users/views/test_user_claim.py (hunk 3 above)
Comment (mfraezz) [reply api_tests/users/views/test_user_claim.py]: Using this as a fixture can also capture "Welcome"-type mails sent at user creation, but using it as the last fixture avoids this problem

File: api/users/permissions.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (sloria) [reply api/users/permissions.py]: Oh right, I was misinterpreting this.

File: api/users/views.py (hunk 1 above)
Comment (sloria) [reply api/users/views.py]: I think the URL map should always be initialized when running the app since we call `init_app` in `manage.py`. Not sure about tests, though. In any case, I don't have a problem keeping this here.

File: api_tests/users/views/test_user_claim.py (hunk 2 above)
Comment (sloria) [reply api_tests/users/views/test_user_claim.py]: I don't feel strongly about removing it. As suggested by Corgibytes, it'd be good for us to be more judicious about adding integration tests, since they're relatively expensive (due to all the db work they need).

Perhaps we could come up with a way to test this without any db work.

```python
assert only_supports_methods(ClaimUser, ["POST"])
```

Certainly not a blocker to merging this, though.

File: api/users/views.py (hunk 1 above)
Comment (mfraezz) [reply api/users/views.py]: We call `init_app` with `routes=False` in `manage.py`, which causes it to not call `make_url_map`.

Comment (sloria) [reply api/users/views.py]: Ah ok, good catch.

File: api_tests/users/views/test_user_claim.py (hunk 2 above)
Comment (mfraezz) [reply api_tests/users/views/test_user_claim.py]: It's a pretty simple util, I'll add it

File: api/users/permissions.py (hunk 5 above)
Comment (mfraezz) [reply api/users/permissions.py]: Update: Added a check in the view to `raise ValidationError('Cannot claim disabled account.')`
